modules/fx/arrow-fx/src/test/kotlin/arrow/fx/QueueTest.kt

```diff
@@ -0,0 +1,217 @@
+package arrow.fx
+
+import arrow.core.extensions.list.traverse.traverse
+import arrow.core.fix
+import arrow.core.Tuple3
+import arrow.core.Tuple2
+import arrow.core.Left
+import arrow.core.None
+import arrow.fx.extensions.fx
+import arrow.fx.extensions.io.applicative.applicative
+import arrow.fx.extensions.io.concurrent.concurrent
+import arrow.fx.typeclasses.milliseconds
+import arrow.test.UnitSpec
+import arrow.test.generators.tuple2
+import arrow.test.generators.tuple3
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import kotlinx.coroutines.Dispatchers
+import kotlin.coroutines.CoroutineContext
+
+class QueueTest : UnitSpec() {
+
+  init {
+
+    fun tests(
+      label: String,
+      ctx: CoroutineContext = Dispatchers.Default,
+      queue: (Int) -> IO<Queue<ForIO, Int>>
+    ) {
+
+      "$label - make a queue the add values then retrieve in the same order" {
+        forAll(Gen.list(Gen.int())) { l ->
+          IO.fx {
+            val q = !queue(l.size)
+            !l.traverse(IO.applicative(), q::offer)
+            val nl = !(1..l.size).toList().traverse(IO.applicative()) { q.take() }
+            nl.fix()
+          }.unsafeRunSync() == l
+        }
+      }
+
+      "$label - offer and take a number of values in the same order" {
+        forAll(Gen.tuple3(Gen.int(), Gen.int(), Gen.int())) { t ->
+          IO.fx {
+            val q = !queue(3)
+            !q.offer(t.a)
+            !q.offer(t.b)
+            !q.offer(t.c)
+            val first = !q.take()
+            val second = !q.take()
+            val third = !q.take()
+            Tuple3(first, second, third)
+          }.unsafeRunSync() == t
+        }
+      }
+
+      "$label - time out taking from an empty queue" {
```

Comment:
This should be rewritten to not rely on `unsafeRunTimed`. `unsafeRunTimed` works in this case because of implementation details. It can be rewritten in a fully pure way, see snippet below.

```
"$label - time out taking from an empty queue" {
  IO.fx {
    val wontComplete = queue(10).flatMap(Queue<ForIO, Int>::take)
    val start = !effect { System.currentTimeMillis() }
    val received = !wontComplete
      .map { Some(it) }
      .waitFor(100.milliseconds, default = just(None))
    val elapsed = !effect { System.currentTimeMillis() - start }
    !effect { received shouldBe None }
    !effect { (elapsed >= 100) shouldBe true }
  }.unsafeRunSync()
}
```

Note: `waitFor` raises an error instead of lifting to `Option`.